Use Flyway for DB-based resource groups migrations

posulliv opened the pull request:
Addresses schema migration for DB-based resource group implementation - #4191

aalbu commented on plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/FlywayMigration.java:
```diff
@@ -0,0 +1,45 @@
+        flyway = Flyway.configure()
+                .dataSource(config.getConfigDbUrl(), config.getConfigDbUser(), config.getConfigDbPassword())
+                .locations("/db/migration/mysql")
+                .baselineOnMigrate(true)
```
```suggestion
                .baselineOnMigrate(true)
                .baselineVersion("0")
```
When migrating against a non-empty schema with no schema history table, the schema is baselined with version 1, by default.  Since your migration (`V1__base_resource_groups_schema.sql`) has version 1, it would get skipped, in that scenario.

electrum commented on plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/DbResourceGroupsModule.java:
```diff
@@ -28,6 +28,8 @@
     public void configure(Binder binder)
     {
         configBinder(binder).bindConfig(DbResourceGroupConfig.class);
+        binder.bind(FlywayMigration.class).in(Scopes.SINGLETON);
```
Actually, this should be injected into the DAO provider to ensure the migration is done before the database is accessed.

electrum commented on plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/ResourceGroupsDao.java:
```diff
@@ -106,4 +57,115 @@ String getExactMatchResourceGroup(
             @Bind("source") String source,
             @Bind("query_type") String queryType);
+
+    @SqlUpdate("INSERT INTO resource_groups_global_properties\n" +
```
Why all these changes? It seems unrelated to Flyway. At a minimum, please move to a separate commit that explains the motivation.

posulliv replied:
I made these changes so I could get tests running using the MySQL DAO instead of the H2 one. I backed this change out though for now.

electrum reviewed (commented):
How would someone who already has a MySQL database created with the old version get it synced to Flyway? Ideally, we could do that automatically, but if not, we need to document it.

posulliv commented:
@electrum I updated the SQL in the initial migration to be a `CREATE TABLE IF NOT EXISTS` to cover the case of users that already have the resource groups tables created in MySQL. I also added a test to attempt to cover this by manually creating some of the tables in the old schema and then doing a migration.

aalbu commented on plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/MysqlDaoProvider.java:
```diff
@@ -28,11 +28,14 @@
+        requireNonNull(migration, "migration is null");
         requireNonNull(config, "config is null");
         MysqlDataSource dataSource = new MysqlDataSource();
         dataSource.setURL(requireNonNull(config.getConfigDbUrl(), "resource-groups.config-db-url is null"));
```
We always require this property to be non-null, so we should annotate it win `@NotNull` and get rid of the `requireNonNull()` calls.  Up to you, but if you decide to make that change, please do it in a separate commit.

posulliv replied:
I am going to make a follow up PR to add migrations for Oracle once this goes in. I will make this change in the follow up PR.

aalbu commented on plugin/trino-resource-group-managers/src/test/java/io/trino/plugin/resourcegroups/db/TestDbResourceGroupsFlywayMigration.java:
```diff
@@ -0,0 +1,118 @@
+                .withUsername("test")
+                .withPassword("test");
+        mysqlServer.start();
+        createOldSchema();
```
I think that for completeness, we should have tests covering all pre-existing conditions:

- empty database
- database with the old schema present
- database without the old schema, but with some other random table

Flyway behaves differently in each of these scenarios.

posulliv replied:
Good idea. Updated to add tests for these conditions.

posulliv commented:
@aalbu thanks for having a look! Updated with your suggestions.

electrum commented on plugin/trino-resource-group-managers/src/main/java/io/trino/plugin/resourcegroups/db/DbResourceGroupsModule.java:
```diff
@@ -33,4 +35,11 @@ public void configure(Binder binder)
+
+    @Singleton
+    @Provides
+    public FlywayMigration flywayMigration(DbResourceGroupConfig config)
```
This can use normal injection. Just add `@Inject` on the constructor and bind it as a singleton.

electrum replied:
What we should add here is something to ensure the database is migrated before it is used. To do this create an empty interface:
```java
@SuppressWarnings("MarkerInterface")
public interface DatabaseReady {}
```
Then make `FlywayMigration` implement it and bind it directly and to the interface:
```java
binder.bind(FlywayMigration.class).asEagerSingleton();
binder.bind(DatabaseReady.class).to(FlywayMigration.class).in(SINGLETON);
```
Finally, add it as a constructor parameter to the places where the database is used:
```java
@Inject
public MysqlDaoProvider(DbResourceGroupConfig config, @SuppressWarnings("unused") DatabaseReady ignored)
{ ... }
```
As part of this, I would actually change `MysqlDaoProvider` to take a `Jdbi` and create it once here, in the module, so that all the usages of `DatabaseReady` are together:
```java
@Provides
@Singleton
public static Jdbi create(DbResourceGroupConfig config, @SuppressWarnings("unused") DatabaseReady ignored)
{
    return Jdbi.create(config.getConfigDbUrl(), config.getConfigDbUser(), config.getConfigDbPassword());
}
```

posulliv commented:
@electrum thank you for the detailed comments! I applied your suggestions and this is ready for another look whenever you get a chance.

posulliv commented:
@electrum sorry to bug you again but this is ready for another review whenever you get a chance. thanks!!

electrum commented:
We recently found that doing migrations as part of Guice injection with the `DatabaseReady` pattern doesn't work well when there are failures, since Guice's handling of failures during injection are not good for this use case. Even though we bind `FlywayMigration` as a singleton, if its construction fails, Guice will try to create it again for each singleton it is injected into. This means we will attempt to run the migrations multiple times, with confusing log messages and the typical terrible "Unable to create injector" exception with many repeated errors.

Instead, it's better to run Flyway manually before invoking `Bootstrap` in `DbResourceGroupConfigurationManagerFactory`:
```java
log.info("Performing migrations...");
DbResourceGroupConfig db = new ConfigurationFactory(config).build(DbResourceGroupConfig.class);
Flyway flyway = Flyway.configure()
        .dataSource(db.getConfigDbUrl(), db.getConfigDbUser(), db.getConfigDbPassword())
        ...
MigrateResult result = flyway.migrate();
log.info("Performed %s migrations", result.migrationsExecuted);
```
So the `DatabaseReady` stuff is actually not needed with this approach.
